### PR TITLE
feat(guilds): add roles & invites managers

### PR DESCRIPTION
**Pull request by Amirust**

### CODE REVIEW

**Comment by Amirust** [comment]
Pls mrg dev in2 master asap!!!!!!

File: src/api/entities/guild/Guild.ts [hunk 1]
```diff
@@ -89,12 +105,24 @@ export class Guild extends AbstractViewableGuild {
     return this.systemChannelId ? this.channels.cache.get<AnyGuildWritableChannel>(this.systemChannelId, options) : undefined
   }
 
+  async edit(data: Partial<RawGuildData>): Promise<this | undefined> {
```
Comment (Mirdukkk) [inline src/api/entities/guild/Guild.ts]: should use GuildEditData

File: src/api/entities/guild/Guild.ts [hunk 2]
```diff
@@ -89,12 +105,24 @@ export class Guild extends AbstractViewableGuild {
...
+    const success = await this.app.internals.actions.editGuild(this.id, data)
```
Comment (Mirdukkk) [inline src/api/entities/guild/Guild.ts]: should use app.guilds.create

File: src/api/entities/guild/Guild.ts [hunk 3]
```diff
@@ -89,12 +105,24 @@ export class Guild extends AbstractViewableGuild {
...
+  }
+
   async leave(): Promise<this | undefined> {
     const success = await this.app.guilds.leave(this.id)
 
     return success ? this : undefined
   }
 
+  async delete(): Promise<this | undefined> {
+    const success = await this.app.internals.actions.deleteGuild(this.id)
```
Comment (Mirdukkk) [inline src/api/entities/guild/Guild.ts]: should use app.guilds.delete

File: src/api/entities/guild/interfaces/RawGuildCreateData.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  region?: string
+  icon?: Buffer | ArrayBuffer
+  verificationLevel?: number
+  defaultNotifications?: number
+  explicitContentFilter?: number
+  roles?: RawRoleCreateData[]
+  channels?: RawGuildChannelCreateData[]
+  afkChannelId?: string
+  afkTimeout?: number
+  systemChannelId?: string
```
Comment (Mirdukkk) [inline src/api/entities/guild/interfaces/RawGuildCreateData.ts]: should use snake_case

File: src/api/managers/ApplicationGuildsManager.ts [hunk 5]
```diff
@@ -19,6 +21,72 @@ export class ApplicationGuildsManager extends EntitiesManager {
     })
   }
 
+  async create(data: GuildCreateData): Promise<Guild | undefined> {
```
Comment (Mirdukkk) [inline src/api/managers/ApplicationGuildsManager.ts]: GuildCreateData OR RawGuildCreateData

File: src/api/managers/ApplicationGuildsManager.ts [hunk 6]
```diff
@@ -19,6 +21,72 @@ export class ApplicationGuildsManager extends EntitiesManager {
...
+      name: data.name,
+    }
+
+    attach(payload, data, {
+      props: [
+        'icon',
+        [ 'verificationLevel', 'verification_level' ],
+        [ 'defaultNotifications', 'default_message_notifications' ],
+        [ 'explicitContentFilter', 'explicit_content_filter' ],
+        'roles',
```
Comment (Mirdukkk) [inline src/api/managers/ApplicationGuildsManager.ts]: should resolve roles to raw when using GuildCreateData

File: src/api/managers/ApplicationGuildsManager.ts [hunk 7]
```diff
@@ -19,6 +21,72 @@ export class ApplicationGuildsManager extends EntitiesManager {
...
+        'channels',
```
Comment (Mirdukkk) [inline src/api/managers/ApplicationGuildsManager.ts]: should resolve channels to raw when using GuildCreateData

File: src/api/managers/invites/InviteGuildsManager.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
+import { AnyInvitableChannelResolvable, EntitiesCacheManager, EntitiesUtil, Invite, InviteCreateOptions } from '@src/api'
+import { EntitiesManager } from '@src/api/managers/EntitiesManager'
+import { DiscordRestApplication } from '@src/core'
+import { Keyspaces } from '@src/constants'
+import { InviteGuildsManagerData } from '@src/api/managers/invites/InviteGuildsManagerData'
+import { DiscordooError, resolveChannelId, resolveGuildId } from '@src/utils'
+
+export class InviteGuildsManager extends EntitiesManager {
+  cache: EntitiesCacheManager<Invite>
```
Comment (Mirdukkk) [inline src/api/managers/invites/InviteGuildsManager.ts]: public cache

File: src/api/managers/invites/InviteGuildsManager.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    if (!channelId) throw new DiscordooError('InviteGuildsManager#create', 'Cannot create invite without channel id.')
+
+    const response = await this.app.internals.actions.createInvite(channelId, {
+      max_age: options.maxAge,
+      max_uses: options.maxUses,
+      temporary: options.temporary,
+      unique: options.unique,
+      target_type: options.targetType,
+      target_user_id: options.targetUserId,
+      target_application_id: options.targetApplicationId
```
Comment (Mirdukkk) [inline src/api/managers/invites/InviteGuildsManager.ts]: use attach

File: src/api/managers/invites/InviteGuildsManager.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      const invite = await new Invite(this.app).init(response.result)
+
+      await this.cache.set(invite.code, invite)
+
+      return invite
+    }
+  }
+
+  async delete(inviteCode: string, reason?: string): Promise<boolean> {
+    const response = await this.app.internals.actions.deleteInvite(inviteCode, reason)
```
Comment (Mirdukkk) [inline src/api/managers/invites/InviteGuildsManager.ts]: InviteResolvable & 'cannot delete invite without code'

File: src/api/managers/roles/GuildRolesManager.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    this.cache = new EntitiesCacheManager<Role>(this.app, {
+      keyspace: 'roles',
+      storage: this.guildId,
+      entity: 'Role',
+      policy: 'roles'
+    })
+  }
+
+  async fetch(id: string): Promise<Role | undefined> {
+    return await this.app.roles.fetchOne(this.guildId, id)
```
Comment (Mirdukkk) [inline src/api/managers/roles/GuildRolesManager.ts]: no need to use async/await here

File: src/api/managers/roles/GuildRolesManager.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  async create(data: RoleCreateData | RawRoleCreateData, reason?: string): Promise<Role | undefined> {
+    const role = await this.app.roles.create(this.guildId, data, reason)
+    if (role) await this.cache.set(role.id, role)
```
Comment (Mirdukkk) [inline src/api/managers/roles/GuildRolesManager.ts]: cache will be updated automatically. no need to update it here

File: src/api/managers/roles/GuildRolesManager.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    return role
+  }
+
+  async edit(id: string, data: RoleCreateData | RawRoleCreateData): Promise<Role | undefined> {
+    const updated = await this.app.roles.edit(this.guildId, id, data)
+    if (updated) await this.cache.set(updated.id, updated)
```
Comment (Mirdukkk) [inline src/api/managers/roles/GuildRolesManager.ts]: cache will be updated automatically. no need to update it here

File: src/api/managers/roles/GuildRolesManager.ts [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    return updated
+  }
+
+  async delete(id: string, reason?: string): Promise<boolean> {
+    const result = await this.app.roles.delete(this.guildId, id, reason)
+    if (result) await this.cache.delete(id)
```
Comment (Mirdukkk) [inline src/api/managers/roles/GuildRolesManager.ts]: cache will be updated automatically. no need to update it here

File: src/core/apps/actions/RestApplicationActions.ts [hunk 15]
```diff
@@ -112,11 +113,11 @@ export class RestApplicationActions {
       .post<RawAppCommandData>()
   }
 
-  createGuild(name: string, data: any /* TODO: GuildCreateData */) {
+  createGuild(data: RawGuildCreateData) {
```
Comment (Mirdukkk) [inline src/core/apps/actions/RestApplicationActions.ts:117]: add return type

File: src/api/managers/roles/GuildRolesManager.ts [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+    this.cache = new EntitiesCacheManager<Role>(this.app, {
+      keyspace: 'roles',
+      storage: this.guildId,
+      entity: 'Role',
+      policy: 'roles'
+    })
+  }
+
+  async fetch(id: string): Promise<Role | undefined> {
```
Comment (Mirdukkk) [inline src/api/managers/roles/GuildRolesManager.ts]: fetch() should automatically decide whether to use fetchOne or fetchMany. also no fetchMany method is present

**Review by Mirdukkk — CHANGES_REQUESTED** [review]
also, no classes and interfaces are exported. they need to be moved to local indexes in the folders in which they are located. you also need to create an ApplicationInvitesManager, and initialize it in DiscordRestApplication.

File: src/api/managers/invites/InviteGuildsManager.ts [hunk 17]
```diff
@@ -0,0 +1,19 @@
+import { EntitiesCacheManager, InviteGuild } from '@src/api'
+import { EntitiesManager } from '@src/api/managers/EntitiesManager'
+import { DiscordRestApplication } from '@src/core'
+import { Keyspaces } from '@src/constants'
+
+export class InviteGuildsManager extends EntitiesManager {
+  cache: EntitiesCacheManager<InviteGuild>
```
Comment (Mirdukkk) [inline src/api/managers/invites/InviteGuildsManager.ts]:
```suggestion
  public cache: EntitiesCacheManager<InviteGuild>
```

File: src/api/entities/guild/Guild.ts [hunk 18]
```diff
@@ -89,12 +110,129 @@ export class Guild extends AbstractViewableGuild {
...
+    return this.edit({ banner }, reason)
+  }
+
+  async setWidgetEnabled(enabled: boolean, reason?: string): Promise<this | undefined> {
+    return this.edit({ widgetEnabled: enabled }, reason)
+  }
+
+  async setWidgetChannel(widgetChannel: GuildChannelResolvable, reason?: string): Promise<this | undefined> {
+    const widgetChannelId = resolveChannelId(widgetChannel)
+    if (!widgetChannelId) throw new DiscordooError('Guild#setWidgetChannel', 'Cannot set widget channel to unknown channel.')
```
Comment (Mirdukkk) [inline src/api/entities/guild/Guild.ts]: please rename all errors in this file to the general format: "Cannot do X without id."

**Review by Mirdukkk — CHANGES_REQUESTED** [review]

File: src/api/entities/guild/interfaces/GuildEditData.ts [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+  features?: GuildFeatures[] //
+  description?: string //
+  banner?: BufferResolvable //
+  splash?: BufferResolvable //
+  discoverySplash?: BufferResolvable
+  ownerId?: string //
+  widgetEnabled?: boolean
+  widgetChannelId?: string
+  premiumProgressEnabled?: boolean //
+  safetyAlertsChannelId?: string //
```
Comment (Mirdukkk) [inline src/api/entities/guild/interfaces/GuildEditData.ts]:
```suggestion
  name?: string
  region?: string
  icon?: BufferResolvable
  verificationLevel?: GuildVerificationLevels
  defaultNotifications?: GuildDefaultMessageNotificationLevel
  explicitContentFilter?: GuildExplicitContentFilterLevels
  afkChannelId?: string
  afkTimeout?: number
  systemChannelId?: string
  systemChannelFlags?: number
  rulesChannelId?: string
  publicUpdatesChannelId?: string
  preferredLocale?: DiscordLocale
  features?: GuildFeatures[]
  description?: string
  banner?: BufferResolvable
  splash?: BufferResolvable
  discoverySplash?: BufferResolvable
  ownerId?: string
  widgetEnabled?: boolean
  widgetChannelId?: string
  premiumProgressEnabled?: boolean
  safetyAlertsChannelId?: string
```

**Review by Mirdukkk — APPROVED** [review]

**Pull request merged**